### PR TITLE
Playlist creation modded into a threading model v2

### PR DESCRIPTION
**Pull request by quickmic**
## Description
VIDEO:
https://1drv.ms/u/s!AsC6yIANTPi5euEisC9FrfzSEIg

btw, the interruption happened, cause of a breakpoint I forgot to remove

## Motivation and Context
Poor playlist Performance on large database

## How Has This Been Tested?
Eclipse X86 (amd64) Linux Mint

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by ksooo — CHANGES_REQUESTED** [review]
Could you please adapt your changes to match or coding guidelines - esp. naming of variables, parameters and member variables. Thanks.

File: xbmc/windows/GUIMediaWindow.cpp [hunk 1]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
  *
  * This function is called by OnClick()
  */
+class QueueMediaAsync : public CJob
+{
+public:
+  QueueMediaAsync(int iPlaylist, std::unique_ptr<CFileItemList> Localm_vecItems)
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: const std...&, please. Here and at other places.

File: xbmc/windows/GUIMediaWindow.cpp [hunk 2]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+  {
+    m_iPlaylist = iPlaylist;
+    m_Localm_vecItems.reset(new CFileItemList);
+    m_Localm_vecItems->Copy(*Localm_vecItems);
+  }
+
+  ~QueueMediaAsync(void) {}
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: remove this

File: xbmc/windows/GUIMediaWindow.cpp [hunk 3]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+      // Perform Playlist refresh every 50 items
+      if ( i % 50 == 0 )
+      {
+        CGUIMessage msg(GUI_MSG_PLAYLIST_CHANGED, 0, 0);
+        CServiceBroker::GetGUI()->GetWindowManager().SendMessage(msg);
+        Event.WaitMSec(100);
+      }
+
+      CFileItemPtr nItem = m_Localm_vecItems->Get(i);
+      PlaylistSize = CServiceBroker::GetPlaylistPlayer().GetPlaylist(m_iPlaylist).size();
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: Move this  outside the loop. No need to obtain the same value on each iteration.

File: xbmc/windows/GUIMediaWindow.cpp [hunk 4]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+
+  ~QueueMediaAsync(void) {}
+
+  bool DoWork(void)
+  {
+    bool ItemDetected = false;
+    int PlaylistSize = 0;
+    int Localm_vecItemsSize = m_Localm_vecItems->Size();
+
+    for ( int i = 0; i < Localm_vecItemsSize; i++ )
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: No blanks after opening bracket and before closing bracket. Here and at other places.

File: xbmc/windows/GUIMediaWindow.cpp [hunk 5]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+
+      // Perform Playlist refresh every 50 items
+      if ( i % 50 == 0 )
+      {
+        CGUIMessage msg(GUI_MSG_PLAYLIST_CHANGED, 0, 0);
+        CServiceBroker::GetGUI()->GetWindowManager().SendMessage(msg);
+        Event.WaitMSec(100);
+      }
+
+      CFileItemPtr nItem = m_Localm_vecItems->Get(i);
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: const CFile...

variable name violates coding guidelines

File: xbmc/windows/GUIMediaWindow.cpp [hunk 6]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+      PlaylistSize = CServiceBroker::GetPlaylistPlayer().GetPlaylist(m_iPlaylist).size();
+
+      //Verify if already added in "OnPlayAndQueueMedia". If that's the case, it must be at the end of the playlist, so only verify from index i to PlaylistSize
+      for ( int j = i; j < PlaylistSize; j++ )
+      {
+        CFileItemPtr pItem = CServiceBroker::GetPlaylistPlayer().GetPlaylist(m_iPlaylist).GetItem(j);
+
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: const CFile...

No need to obtain playlist on each iteration. Do this once outside the loop.

File: xbmc/windows/GUIMediaWindow.cpp [hunk 7]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+    //Final playlist refresh
+    CGUIMessage msg(GUI_MSG_PLAYLIST_CHANGED, 0, 0);
+    CServiceBroker::GetGUI()->GetWindowManager().SendMessage(msg);
+    return true;
+  }
+  private:
+    int m_iPlaylist = -1;
+    std::unique_ptr<CFileItemList> m_Localm_vecItems;
+    CEvent Event;
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: We prefix member vars with m_

first letter after m_ has to be lower case

File: xbmc/windows/GUIMediaWindow.cpp [hunk 8]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+    // Generate a signature for the itemlist
+    float ItemListContentSignatureCompare = 0;
+
+    if (m_vecItemsSize > 2)
+    {
+      std::string path1 = m_vecItems->Get(0)->GetPath();
+      std::string path2 = m_vecItems->Get(m_vecItemsSize - 1)->GetPath();
+      std::size_t h1 = std::hash<std::string>{}(path1);
+      std::size_t h2 = std::hash<std::string>{}(path2);
+      ItemListContentSignatureCompare = h1 + h2 + m_vecItemsSize;
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: All 4 vars can be const

File: xbmc/windows/GUIMediaWindow.cpp [hunk 9]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+    }
+
+    // Do not reload complete Playlist, if nothing has changed
+    if (ItemListContentSignature == ItemListContentSignatureCompare)
+    { // Skip reloading Playlist (identical playlist according to verification), just play Song according to selection
+      int iPlaylist = m_guiState->GetPlaylist();
+
+      for ( int i = 0; i < m_vecItems->Size(); i++ )
+      {
+        CFileItemPtr nItem = m_vecItems->Get(i);
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: const CFile...

File: xbmc/windows/GUIMediaWindow.cpp [hunk 10]
```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
...
+
+      	  //Only add item if not already covered by "QueueMediaAsync"
+      	  if (i > PlaylistSize)
+      	  {
+      		//i > PlaylistSize means item not added via thread, but maybe already added by this code on a previous run. So we check the playlist
+      		ItemDetected = false;
+
+      		for ( int j = 0; j < PlaylistSize; j++ )
+      		{
+      			CFileItemPtr pItem = CServiceBroker::GetPlaylistPlayer().GetPlaylist(iPlaylist).GetItem(j);
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.cpp]: Move obtaining Playlist outside of the loop.

File: xbmc/windows/GUIMediaWindow.h [hunk 11]
```diff
@@ -186,6 +187,12 @@ class CGUIMediaWindow : public CGUIWindow
   CDirectoryHistory m_history;
   std::unique_ptr<CGUIViewState> m_guiState;
   std::atomic_bool m_vecItemsUpdating = {false};
+
+  //PlaylistAsync Threading
+  float ItemListContentSignature = -1;
+  int QueueMediaAsyncJobID = -1;
```
Comment (ksooo) [inline xbmc/windows/GUIMediaWindow.h]: m_ followed by lower case letter...

**Review by ksooo — CHANGES_REQUESTED** [review]
This needs a bit more love, imo

**Comment by quickmic** [comment]
Thanks for review, I fix the issues asap.

File: xbmc/windows/GUIMediaWindow.cpp (hunk 3 above)
Comment (quickmic) [reply xbmc/windows/GUIMediaWindow.cpp]: The PlaylistSize could have changed in worker thread, cause the OnPlayAndQueueMedia routine could have added an item while worker is active.
e.g. whole "itemlist" has 1000 songs (in my video I use the genre selection as "itemlist")
Now I select one e.g. 10th song from Itemlist. This song is added immediately to playlist. Now worker triggers... 
first loop is "for ( int i = 0; i < Localm_vecItemsSize; i++ )"
i=0 so try adding first song but verify if not already added (10th song already added)
get 
PlaylistSize = CServiceBroker::GetPlaylistPlayer().GetPlaylist(m_iPlaylist).size();
second loop for verification:
for ( int j = i; j < PlaylistSize; j++ )
range 0 to 0 (one item (the 10th song) already in playlist so we check if item = playlistitem
and so on.

Now imagine following. we add 900th song. Worker starts and progress takes time.
On large playlists I can add another song  e.g. 950 via "OnPlayAndQueueMedia" (therefore is the signature verification).
now, e.g. worker is at position 500 but playlistsize is now 502. -> Playlistsize changed.
and so on...

**Comment by ccope** [comment]
You don't need to close the PR every time you need to make more changes, it makes it harder to follow the discussion. Just push more commits to your branch.

**Comment by quickmic** [comment]
Yes I know and I tried to modify the PR, but once again I f*** it up as you can see.
btw, I think I know were I went wrong, but to be honest, I don't wanna fiddle around with fixing broken repos. User-friendlyness isn't the strong side of GIT.
fyi, I used the instructions (https://github.com/xbmc/xbmc/blob/master/docs/GIT-FU.md) starting from 8.2
and sorry, that the comments previous comments are lost. As mentioned, it was not on purpose.

**Comment by ccope** [comment]
No worries. Looking at that section, I think those instructions are likely to cause merge conflicts. Just keep pushing new commits until your code is in good shape, then you can squash them into a single commit with git rebase

**Pull request closed without merging**